File: tests/cpp/unit/ColorHelperTests.cpp
#include "services/ColorHelper.h"

#include <gtest/gtest.h>

using namespace imeindicator;

namespace {

void expectChannelNear(float actual, float expected)
{
    // 整数 RGB を float に正規化した値の比較は誤差 1/255 以内であれば一致とみなす。
    EXPECT_NEAR(actual, expected, 1.0f / 255.0f / 2.0f);
}

void expectColor(const D2D1_COLOR_F& c, float r, float g, float b, float a)
{
    expectChannelNear(c.r, r);
    expectChannelNear(c.g, g);
    expectChannelNear(c.b, b);
    expectChannelNear(c.a, a);
}

} // namespace

// ============================================================
// parseColor — #RRGGBB 形式
// ============================================================

TEST(ColorHelperTests, ParseColor_RRGGBB_Red)
{
    expectColor(services::ColorHelper::parseColor("#FF0000"), 1.0f, 0.0f, 0.0f, 1.0f);
}
TEST(ColorHelperTests, ParseColor_RRGGBB_Green)
{
    expectColor(services::ColorHelper::parseColor("#00FF00"), 0.0f, 1.0f, 0.0f, 1.0f);
}
TEST(ColorHelperTests, ParseColor_RRGGBB_Blue)
{
    expectColor(services::ColorHelper::parseColor("#0000FF"), 0.0f, 0.0f, 1.0f, 1.0f);
}
TEST(ColorHelperTests, ParseColor_RRGGBB_White)
{
    expectColor(services::ColorHelper::parseColor("#FFFFFF"), 1.0f, 1.0f, 1.0f, 1.0f);
}
TEST(ColorHelperTests, ParseColor_RRGGBB_Black)
{
    expectColor(services::ColorHelper::parseColor("#000000"), 0.0f, 0.0f, 0.0f, 1.0f);
}

// ============================================================
// parseColor — #RGB 形式
// ============================================================

TEST(ColorHelperTests, ParseColor_RGB_Red)
{
    expectColor(services::ColorHelper::parseColor("#F00"), 1.0f, 0.0f, 0.0f, 1.0f);
}
TEST(ColorHelperTests, ParseColor_RGB_Green)
{
    expectColor(services::ColorHelper::parseColor("#0F0"), 0.0f, 1.0f, 0.0f, 1.0f);
}

// ============================================================
// parseColor — #AARRGGBB 形式
// ============================================================

TEST(ColorHelperTests, ParseColor_AARRGGBB_SemiTransparentRed)
{
    auto c = services::ColorHelper::parseColor("#80FF0000");
    expectChannelNear(c.a, 0x80 / 255.0f);
    expectChannelNear(c.r, 1.0f);
    expectChannelNear(c.g, 0.0f);
    expectChannelNear(c.b, 0.0f);
}

TEST(ColorHelperTests, ParseColor_AARRGGBB_OpaqueWhite)
{
    auto c = services::ColorHelper::parseColor("#FFFFFFFF");
    expectColor(c, 1.0f, 1.0f, 1.0f, 1.0f);
}

// ============================================================
// parseColor — 不正入力 → 灰色
// ============================================================

TEST(ColorHelperTests, ParseColor_Empty_ReturnsGray)
{
    expectColor(services::ColorHelper::parseColor(""), 0.5f, 0.5f, 0.5f, 1.0f);
}
TEST(ColorHelperTests, ParseColor_WhitespaceOnly_ReturnsGray)
{
    expectColor(services::ColorHelper::parseColor("   "), 0.5f, 0.5f, 0.5f, 1.0f);
}
TEST(ColorHelperTests, ParseColor_InvalidChars_ReturnsGray)
{
    expectColor(services::ColorHelper::parseColor("#ZZZZZZ"), 0.5f, 0.5f, 0.5f, 1.0f);
}
TEST(ColorHelperTests, ParseColor_InvalidLength_ReturnsGray)
{
    expectColor(services::ColorHelper::parseColor("#12345"), 0.5f, 0.5f, 0.5f, 1.0f);
}

// ============================================================
// toRgbHex / toArgbHex
// ============================================================

TEST(ColorHelperTests, ToRgbHex_Black)
{
    EXPECT_EQ(services::ColorHelper::toRgbHex({0.0f, 0.0f, 0.0f, 1.0f}), "#000000");
}
TEST(ColorHelperTests, ToRgbHex_White)
{
    EXPECT_EQ(services::ColorHelper::toRgbHex({1.0f, 1.0f, 1.0f, 1.0f}), "#FFFFFF");
}
TEST(ColorHelperTests, ToRgbHex_RoundTrip)
{
    auto orig = services::ColorHelper::parseColor("#3B82F6");
    auto hex = services::ColorHelper::toRgbHex(orig);
    auto parsed = services::ColorHelper::parseColor(hex);
    EXPECT_EQ(hex, "#3B82F6");
    expectChannelNear(orig.r, parsed.r);
    expectChannelNear(orig.g, parsed.g);
    expectChannelNear(orig.b, parsed.b);
}
TEST(ColorHelperTests, ToArgbHex_RoundTrip)
{
    auto orig = services::ColorHelper::parseColor("#80FF0000");
    auto hex = services::ColorHelper::toArgbHex(orig);
    auto parsed = services::ColorHelper::parseColor(hex);
    EXPECT_EQ(hex, "#80FF0000");
    expectChannelNear(orig.a, parsed.a);
    expectChannelNear(orig.r, parsed.r);
    expectChannelNear(orig.g, parsed.g);
    expectChannelNear(orig.b, parsed.b);
}

// ============================================================
// isValidHexColor
// ============================================================

TEST(ColorHelperTests, IsValidHexColor_Valid)
{
    EXPECT_TRUE(services::ColorHelper::isValidHexColor("#FF0000"));
    EXPECT_TRUE(services::ColorHelper::isValidHexColor("#F00"));
    EXPECT_TRUE(services::ColorHelper::isValidHexColor("#80FF0000"));
    EXPECT_TRUE(services::ColorHelper::isValidHexColor("#F00F"));
    EXPECT_TRUE(services::ColorHelper::isValidHexColor("FF0000"));  // # 無しでも許容
}
TEST(ColorHelperTests, IsValidHexColor_Invalid)
{
    EXPECT_FALSE(services::ColorHelper::isValidHexColor("#ZZZZZZ"));
    EXPECT_FALSE(services::ColorHelper::isValidHexColor("#12345"));
    EXPECT_FALSE(services::ColorHelper::isValidHexColor(""));
    EXPECT_FALSE(services::ColorHelper::isValidHexColor("   "));
}

File: tests/csharp/IMEIndicator.Tests/Services/ColorHelperTests.cs
// Copyright (C) 2026 IMEIndicator Project
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License v2 or later.
// See COPYING in the repository root for the full license text.

using IMEIndicator.Services;

using Xunit;

namespace IMEIndicator.Tests.Services;

/// <summary>
/// <see cref="ColorHelper"/> のテスト。現行 tests/cpp/unit/ColorHelperTests.cpp の移植。
/// C++ 版は内部表現が float（D2D1_COLOR_F、0.0〜1.0）のため比較に誤差許容（1/510）を使っていたが、
/// C# 版は System.Drawing.Color（byte、0〜255）を直接パースし丸め誤差が発生しないため、
/// 本ファイルでは全て厳密一致で検証する。
/// </summary>
public sealed class ColorHelperTests
{
    // 不正入力時のフォールバック色（ColorHelper.GrayFallback と同じ値）。
    private static readonly Color Gray = Color.FromArgb(255, 128, 128, 128);

    // ============================================================
    // ParseColor — #RRGGBB 形式
    // ============================================================

    [Theory]
    [InlineData("#FF0000", 255, 255, 0, 0)]
    [InlineData("#00FF00", 255, 0, 255, 0)]
    [InlineData("#0000FF", 255, 0, 0, 255)]
    [InlineData("#FFFFFF", 255, 255, 255, 255)]
    [InlineData("#000000", 255, 0, 0, 0)]
    public void ParseColor_RRGGBB_MatchesExpectedArgb(string hex, byte a, byte r, byte g, byte b)
    {
        Assert.Equal(Color.FromArgb(a, r, g, b), ColorHelper.ParseColor(hex));
    }

    // ============================================================
    // ParseColor — #RGB 形式
    // ============================================================

    [Theory]
    [InlineData("#F00", 255, 255, 0, 0)]
    [InlineData("#0F0", 255, 0, 255, 0)]
    public void ParseColor_RGB_MatchesExpectedArgb(string hex, byte a, byte r, byte g, byte b)
    {
        Assert.Equal(Color.FromArgb(a, r, g, b), ColorHelper.ParseColor(hex));
    }

    // ============================================================
    // ParseColor — #AARRGGBB 形式
    // ============================================================

    [Fact]
    public void ParseColor_AARRGGBB_SemiTransparentRed()
    {
        var c = ColorHelper.ParseColor("#80FF0000");
        Assert.Equal(0x80, c.A);
        Assert.Equal(255, c.R);
        Assert.Equal(0, c.G);
        Assert.Equal(0, c.B);
    }

    [Fact]
    public void ParseColor_AARRGGBB_OpaqueWhite()
    {
        Assert.Equal(Color.FromArgb(255, 255, 255, 255), ColorHelper.ParseColor("#FFFFFFFF"));
    }

    // ============================================================
    // ParseColor — 不正入力 → 灰色
    // ============================================================

    [Fact]
    public void ParseColor_Null_ReturnsGray()
    {
        // string? を受け取る InlineData(null) は params object[] の単一引数解決が曖昧になりうるため、
        // null ケースのみ専用の Fact に分離する。
        Assert.Equal(Gray, ColorHelper.ParseColor(null));
    }

    [Theory]
    [InlineData("")]
    [InlineData("   ")]
    [InlineData("#ZZZZZZ")]
    [InlineData("#12345")]
    public void ParseColor_InvalidInput_ReturnsGray(string hex)
    {
        Assert.Equal(Gray, ColorHelper.ParseColor(hex));
    }

    // ============================================================
    // ToRgbHex / ToArgbHex
    // ============================================================

    [Fact]
    public void ToRgbHex_Black()
    {
        Assert.Equal("#000000", ColorHelper.ToRgbHex(Color.FromArgb(255, 0, 0, 0)));
    }

    [Fact]
    public void ToRgbHex_White()
    {
        Assert.Equal("#FFFFFF", ColorHelper.ToRgbHex(Color.FromArgb(255, 255, 255, 255)));
    }

    [Fact]
    public void ToRgbHex_RoundTrip()
    {
        var orig = ColorHelper.ParseColor("#3B82F6");
        var hex = ColorHelper.ToRgbHex(orig);
        var parsed = ColorHelper.ParseColor(hex);

        Assert.Equal("#3B82F6", hex);
        Assert.Equal(orig.R, parsed.R);
        Assert.Equal(orig.G, parsed.G);
        Assert.Equal(orig.B, parsed.B);
    }

    [Fact]
    public void ToArgbHex_RoundTrip()
    {
        var orig = ColorHelper.ParseColor("#80FF0000");
        var hex = ColorHelper.ToArgbHex(orig);
        var parsed = ColorHelper.ParseColor(hex);

        Assert.Equal("#80FF0000", hex);
        Assert.Equal(orig.A, parsed.A);
        Assert.Equal(orig.R, parsed.R);
        Assert.Equal(orig.G, parsed.G);
        Assert.Equal(orig.B, parsed.B);
    }

    // ============================================================
    // IsValidHexColor
    // ============================================================

    [Theory]
    [InlineData("#FF0000")]
    [InlineData("#F00")]
    [InlineData("#80FF0000")]
    [InlineData("#F00F")]
    [InlineData("FF0000")] // # 無しでも許容
    public void IsValidHexColor_Valid_ReturnsTrue(string hex)
    {
        Assert.True(ColorHelper.IsValidHexColor(hex));
    }

    [Fact]
    public void IsValidHexColor_Null_ReturnsFalse()
    {
        Assert.False(ColorHelper.IsValidHexColor(null));
    }

    [Theory]
    [InlineData("#ZZZZZZ")]
    [InlineData("#12345")]
    [InlineData("")]
    [InlineData("   ")]
    public void IsValidHexColor_Invalid_ReturnsFalse(string hex)
    {
        Assert.False(ColorHelper.IsValidHexColor(hex));
    }
}
